docs(Tooltip): add a11y docs

thatblindgeye opened the pull request:
Closes #2596

[Tooltip a11y tab](https://patternfly-org-pr-2892-v4.surge.sh/v4/components/tooltip/accessibility)

Current implementation for a11y docs on the tooltip. This is based off of the current changes made in https://github.com/patternfly/patternfly-react/pull/7332 and https://github.com/patternfly/patternfly-react/pull/7335.

Additionally, I added in some verbiage in regards to tooltips on static elements. In https://github.com/patternfly/patternfly-org/issues/1108#issuecomment-495262558 it was mentioned that tooltips on static elements shouldn't be recommended. However, utilizing the pattern of `aria-live` instead of `aria-describedby` does seem to resolve the gaps mentioned.


Original comment:

One proposal I want to put out there is whether we should include verbiage for when a tooltip *shouldn't* be announced to screen reader users. Specifically, if an element already has a label (such as via `aria-label`), and the tooltip contents is essentially the same thing, should the tooltip then be prevented from being announced to screen reader users to avoid redundant/duplicate announcements? Or would the better approach be to avoid adding an `aria-label` if a tooltip is to be attached (as we've seen with using React refs, tooltips may not always be announced, though I've noticed a possible fix could be using `aria-live` on tooltips attached via ref)?

Using the [Tooltip on icon example](https://www.patternfly.org/v4/components/tooltip#on-icon), if the example code was changed so that the button had `aria-label="Copy contents to clipboard"`, a screen reader could announce both that and the tooltip contents, which would be redundant information. This could also be trying to solve a problem that may not really exist or not be a significant issue, though.

patternfly-build commented:
PF4 preview: https://patternfly-org-pr-2892-v4.surge.sh/v4

jessiehuff reviewed (dismissed):
Looks good! Thanks for working on this! 🙂

thatblindgeye commented:
@nicolethoen from the backlog meeting earlier, I noticed one issue regarding Popover and adding hover functionality. After digging through that, based on issue #2871 + conversation in the issue linked in it, do you think it would be better to update the a11y docs for Tooltips taking into account these intended changes to Tooltip + Popover? For the most part may just be a case of re-adding the recommendation to not use tooltips on static elements, and then later adding in any other verbiage the design guidelines end up including.

nicolethoen commented:
> For the most part may just be a case of re-adding the recommendation to not use tooltips on static elements, and then later adding in any other verbiage the design guidelines end up including.

I think I agree - we should include the recommendation of not using tooltips on static elements and then we can follow up if recommendations evolve. 
The only exception being if we already are following the 'yet to be written' recommendations in the examples/demos and then it would just mean that the design guidelines would be one step behind the a11y docs which is not the end of the world.

jessiehuff reviewed (commented):
I really love these updates! I think it's a good call out that you're differentiating labeling from supplementary information. 

I think Nicole makes a good suggestion as well. Adding small snippets/examples helps make the points clearer and more consumable.

thatblindgeye commented:
@nicolethoen @jessiehuff I updated the tooltip doc with the new template. I also added in some code blocks/examples for certain things throughout. For examples with specific props, they're not exactly inline in the table and instead placed in sub-sections, so let me know what you think of that sort of layout or if you have other ideas for how to go about them

jessiehuff reviewed (commented):
I like the new template for this, but I find the top section a little confusing between the React and HTML/CSS bullet points. At first I thought that you meant that everyone needs to add all of those things. For example, I saw the point about adding the role and thought, "isn't that already added for you?" Then I realized that you meant if you're just using the HTML/CSS, not that you need to add HTML/CSS. Admittedly I think previously I just included the React suggestions before and kept the HTML/CSS in the customization section. However, I can understand why it might make sense to keep at the top, but I'd be curious what others think. Do you think it would be helpful for us to separate that into two different categories at the top to keep it clear? Or maybe it was just me that got confused haha.

thatblindgeye commented:
>Or maybe it was just me that got confused haha.

I definitely think you raise a valid point. What do you think of something along the lines of this instead:

```
- Pass in `role="tooltip"` to the element acting as the tooltip component if using the HTML/CSS library
- Pass in `aria="labelledby"` to the tooltip component if using the PatternFly React library, or the `aria-labelledby` attribute to the trigger if using the HTML/CSS library, when the tooltip should act as the primary label for its trigger

or use nested bullets to separate the two libraries:

- Pass in `aria="labelledby"` to the tooltip component if using the PatternFly React library when the tooltip should act as the primary label for its trigger.
    - Pass in the `aria-labelledby` attribute to the trigger if using the HTML/CSS library
```

The "if using X library..." could also be placed first, e.g. "If using the HTML/CSS library, pass in `role="tooltip"`...".

nicolethoen commented:
I think I like your second suggestion more. maybe something like:

To implement an accessible PatternFly tooltip:
  - Avoid using tooltips on static elements such as a div or span, except in cases of truncation.

Using react:
- Pass in aria="labelledby" when the tooltip should act as the primary label for its trigger (with snippet)

Using HTML/CSS:
- Set the aria-labelledby attribute on the trigger when the tooltip should act as the primary label for its trigger. (code snippet)
- Pass in the aria-describedby attribute to the trigger (HTML/CSS) when the tooltip should act as supplementary information. (code snippet)
